src/main/java/team/reborn/energy/api/base/SimpleItemEnergyStorage.java

```diff
@@ -0,0 +1,135 @@
+	private final long capacity;
+	private final long maxInsert, maxExtract;
+
+	public SimpleItemEnergyStorage(ContainerItemContext ctx, long capacity, long maxInsert, long maxExtract) {
+		StoragePreconditions.notNegative(capacity);
+		StoragePreconditions.notNegative(maxInsert);
+		StoragePreconditions.notNegative(maxExtract);
+
+		this.ctx = ctx;
+		this.startingItem = ctx.getItemVariant().getItem();
+		this.capacity = capacity;
+		this.maxInsert = maxInsert;
+		this.maxExtract = maxExtract;
+	}
+
+	/**
+	 * Used to check if the context still has the starting item. (If it doesn't we can't do anything).
+	 */
+	private boolean hasStartingItem() {
+		return ctx.getItemVariant().isOf(startingItem) && ctx.getAmount() > 0;
+	}
+
+	/**
+	 * Try to set the energy of the stack to {@code energyAmountPerCount}, return true if success.
+	 */
+	private boolean trySetEnergy(long energyAmountPerCount, long count, TransactionContext transaction) {
+		ItemStack newStack = ctx.getItemVariant().toStack();
+		SimpleBatteryItem.setStoredEnergyUnchecked(newStack, energyAmountPerCount);
+		ItemVariant newVariant = ItemVariant.of(newStack);
+
+		// Try to convert exactly `count` items.
+		try (Transaction nested = transaction.openNested()) {
+			if (ctx.extract(ctx.getItemVariant(), count, nested) == count && ctx.insert(newVariant, count, nested) == count) {
+				nested.commit();
+				return true;
+			}
+		}
+
+		return false;
+	}
+
+	@Override
+	public boolean supportsInsertion() {
+		return maxInsert > 0 && hasStartingItem();
+	}
+
+	@Override
+	public long insert(long maxAmount, TransactionContext transaction) {
+		StoragePreconditions.notNegative(maxAmount);
+
+		// Make sure the item matches.
+		if (!hasStartingItem()) return 0;
+
+		long count = ctx.getAmount();
+
+		long maxAmountPerCount = maxAmount / count;
+		long currentAmountPerCount = getAmount() / count;
+		long insertedPerCount = Math.min(maxInsert, Math.min(maxAmountPerCount, capacity - currentAmountPerCount));
+
+		if (insertedPerCount > 0) {
+			if (trySetEnergy(currentAmountPerCount + insertedPerCount, count, transaction)) {
+				return insertedPerCount * count;
+			}
+		}
+
+		return 0;
+	}
+
+	@Override
+	public boolean supportsExtraction() {
+		return maxExtract > 0 && hasStartingItem();
+	}
+
+	@Override
+	public long extract(long maxAmount, TransactionContext transaction) {
+		StoragePreconditions.notNegative(maxAmount);
+
+		// Make sure the item matches.
+		if (!hasStartingItem()) return 0;
+
+		long count = ctx.getAmount();
+
+		long maxAmountPerCount = maxAmount / count;
+		long currentAmountPerCount = getAmount() / count;
+		long extractedPerCount = Math.min(maxExtract, Math.min(maxAmountPerCount, currentAmountPerCount));
+
+		if (extractedPerCount > 0) {
+			if (trySetEnergy(currentAmountPerCount - extractedPerCount, count, transaction)) {
+				return extractedPerCount * count;
+			}
+		}
+
+		return 0;
+	}
+
+	@Override
+	public long getAmount() {
+		if (hasStartingItem()) {
```

Comment:
Wait before merging, I'll add something to avoid having to put these checks all over the place.